python/lambdas/copy-files-from-tdr/copy_files_lambda_function_test.py

```diff
@@ -138,5 +156,135 @@ def test_send_message_failure(self, mock_send_message, _, mock_head_object):
+        self.assertEqual(True, result)
+
+
+    @patch('lambda_function.s3_client.get_object')
+    def test_should_raise_an_exception_when_series_does_not_exist(self, mock_get_object):
+        mock_get_object.return_value = {
+            'Body':
+                '{"ConsignmentReference": "TDR-2024-PQXN", "FileReference": "ZDSCFC", '
+                '"TransferInitiatedDatetime": "2024-09-19 07:21:57","UUID": "0000c951-b332-4d45-93e7-8c24eec4b1f1"}'
+             }
+        with self.assertRaises(Exception) as ex:
+            lambda_function.validate_mandatory_fields_exist('any_bucket_patched', 'any_key_patched')
+        self.assertEqual("'Series' is a required property" , str(ex.exception))
+
+    @patch('lambda_function.s3_client.get_object')
+    def test_should_raise_an_exception_when_transfer_initiated_datetime_does_not_exist(self, mock_get_object):
+        mock_get_object.return_value = {
+            'Body':
+                '{"ConsignmentReference": "TDR-2024-PQXN", "FileReference": "ZDSCFC", "Series":"MOCK1 123", '
+                '"UUID": "0000c951-b332-4d45-93e7-8c24eec4b1f1"}'
+             }
+        with self.assertRaises(Exception) as ex:
+            lambda_function.validate_mandatory_fields_exist('any_bucket_patched', 'any_key_patched')
+        self.assertEqual("'TransferInitiatedDatetime' is a required property" , str(ex.exception))
+
+    @patch('lambda_function.s3_client.get_object')
+    def test_should_raise_an_exception_when_UUID_does_not_exist(self, mock_get_object):
+        mock_get_object.return_value = {
+            'Body':
+                '{"ConsignmentReference": "TDR-2024-PQXN", "FileReference": "ZDSCFC", "Series":"MOCK1 123", '
+                '"TransferInitiatedDatetime": "2024-09-19 07:21:57"}'
+             }
+        with self.assertRaises(Exception) as ex:
+            lambda_function.validate_mandatory_fields_exist('any_bucket_patched', 'any_key_patched')
+        self.assertEqual("'UUID' is a required property" , str(ex.exception))
+
+    @patch('lambda_function.s3_client.head_object')
+    def test_should_raise_an_exception_when_the_object_does_not_exist_in_source_bucket(self, mock_head_object):
+        error_response = {
+            'Error': {
+                'Code': '404',
+                'Message': 'Not Found'
+            }
+        }
+
+        client_error = ClientError(error_response, 'HeadObject')
+
+        mock_head_object.side_effect = client_error
+        with self.assertRaises(Exception) as ex:
+            lambda_function.assert_objects_exist('some_bucket', 'some_key')
+
+        self.assertEqual('Object some_key does not exist, underlying error is: '
+                         'An error occurred (404) when calling the HeadObject operation: Not Found' , str(ex.exception))
+
+
+    @patch('lambda_function.s3_client.head_object')
+    def test_should_raise_an_exception_when_the_object_exists_but_metadata_does_not_exist_in_source_bucket(self, mock_head_object):
```

Comment:
I'm not sure that this test name is 100% correct because it says "**when the object exists** but metadata does not exist" but this test does not do anything to verify that the object exists, we only assume that it does because it has reached this stage, but if you were to remove the headObject check in the `assert_objects_exist` method or change the order they were called, this test name wouldn't be true. I think it should just be `test_should_raise_an_exception_when_the_metadata_does_not_exist_in_source_bucket`

Reply:
I disagree, the mock that is setup below, is setup only to throw an exception on metadata, hence it is precisely doing what the name suggests.

Reply:
Yes, sorry if I wasn't clear earlier but what I meant is you have "when_the_object_exists" in the test name but you are not testing that the object exists in this test, you're just testing that the metadata exists

Reply:
Maybe I should have asked "what do you mean by 'object' in your test title"?

Reply:
Calling it a file now.